models/issues/milestone_label.go

```diff
@@ -0,0 +1,318 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package issues
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"code.gitea.io/gitea/models/db"
+	access_model "code.gitea.io/gitea/models/perm/access"
+	user_model "code.gitea.io/gitea/models/user"
+
+	"xorm.io/builder"
+)
+
+// MilestoneLabel represents an milestone-label relation.
+type MilestoneLabel struct {
+	ID          int64 `xorm:"pk autoincr"`
+	MilestoneID int64 `xorm:"UNIQUE(milestoneid_labelid)"`
+	LabelID     int64 `xorm:"UNIQUE(milestoneid_labelid)"`
+}
+
+func init() {
+	db.RegisterModel(new(MilestoneLabel))
+}
+
+// HasMilestoneLabel returns true if milestone has been labeled.
+func HasMilestoneLabel(ctx context.Context, milestoneID, labelID int64) (bool, error) {
+	has, err := db.GetEngine(ctx).Where("milestone_id = ? AND label_id = ?", milestoneID, labelID).Exist(new(MilestoneLabel))
+	if err != nil {
+		return false, err
+	}
+	return has, nil
+}
+
+// GetLabelsByMilestoneID returns all labels that belong to given milestone by ID.
+func GetLabelsByMilestoneID(ctx context.Context, milestoneID int64) ([]*Label, error) {
+	var labels []*Label
+	return labels, db.GetEngine(ctx).Where("milestone_label.milestone_id = ?", milestoneID).
+		Join("LEFT", "milestone_label", "milestone_label.label_id = label.id").
+		Asc("label.name").
+		Find(&labels)
+}
+
+// newMilestoneLabel creates a new label, but it does NOT check if the label is valid for the specified milestone
+// YOU MUST CHECK THIS BEFORE EXECUTING THIS FUNCTION
+func newMilestoneLabel(ctx context.Context, m *Milestone, label *Label, doer *user_model.User) (err error) {
+	if err = db.Insert(ctx, &MilestoneLabel{
+		MilestoneID: m.ID,
+		LabelID:     label.ID,
+	}); err != nil {
+		return err
+	}
+
+	return updateLabelCols(ctx, label, "num_issues", "num_closed_issues", "num_milestones")
+}
+
+// NewMilestoneLabel creates a new milestone-label relation.
+func NewMilestoneLabel(m *Milestone, label *Label, doer *user_model.User) (err error) {
+	hasLabel, err := HasMilestoneLabel(db.DefaultContext, m.ID, label.ID)
+	if hasLabel {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	ctx, committer, err := db.TxContext(db.DefaultContext)
+	if err != nil {
+		return err
+	}
+	defer committer.Close()
+
+	// Do NOT add invalid labels
+	if m.RepoID != label.RepoID && m.Repo.OwnerID != label.OrgID {
+		return nil
+	}
+
+	if err = newMilestoneLabel(ctx, m, label, doer); err != nil {
+		return err
+	}
+
+	m.Labels = nil
+	if err = m.LoadLabels(ctx); err != nil {
+		return err
+	}
```

Comment:
Are there any reasons that we need to clear it and load it from DB instead of adding new labels to `m.Labels` directly? 🤔